### PR TITLE
DO NOT MERGE: feat(rest): reject array values in JSON requests

### PR DESCRIPTION
**Pull request by achrinza**
see #3700

Signed-off-by: Rifa Achrinza <25147899+achrinza@users.noreply.github.com>

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/rest/src/body-parsers/body-parser.json.ts [hunk 1]
```diff
@@ -36,6 +38,9 @@ export class JsonBodyParser implements BodyParser {
 
   async parse(request: Request): Promise<RequestBody> {
     let body = await invokeBodyParserMiddleware(this.jsonParser, request);
+
+    if (!_.isPlainObject(body)) throw RestHttpErrors.invalidRequestBody();
```
Comment (raymondfeng) [inline packages/rest/src/body-parsers/body-parser.json.ts]: Rejecting non-object body unconditionally is not desired. It should be perfect fine to expose an API to accept json array in the body, for example, an array of deals to be added.

**Comment by raymondfeng** [comment]
@achrinza To fix #3700, we probably need to fix the root cause (toObject() is invoked on an array).

**Comment by achrinza** [comment]
@raymondfeng Ah noted; `toObject()` is in `@loopback/repository`. In that case, how should we throw an exception?

**Comment by bajtos** [comment]
Thank you @achrinza for the pull request!

> To fix #3700, we probably need to fix the root cause (toObject() is invoked on an array).

I disagree. IMO, if I decorate my controller method with OpenAPI spec saying that `data` argument allows only objects, then LoopBack's REST layer must ensure my controller method never receives array instances.

In https://github.com/strongloop/loopback-next/issues/3700, I proposed:

> Improve our REST validation to reject Array values for parameters accepting an object.

Similarly, we can reject object (non-array) values for parameters accepting an array.

> Ah noted; toObject() is in @loopback/repository. In that case, how should we throw an exception?

Yes, we should throw an exception with 4xx status code - the client sent us a bad request.

File: packages/rest/src/__tests__/unit/body-parser.unit.ts [hunk 2]
```diff
@@ -90,6 +90,25 @@ describe('body parser', () => {
...
+    const req = givenRequest({
+      url: '/',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      payload: [],
+    });
+    const spec = givenOperationWithRequestBody({
+      description: 'data',
+      content: {},
```
Comment (bajtos) [inline packages/rest/src/__tests__/unit/body-parser.unit.ts]: IMO it's crucial to provide a schema for the request body object, this schema must describe `type: object`.

As I interpret the currently proposed spec, it allows any values for the request body, therefore LoopBack's REST layer must not reject arrays.

File: packages/rest/src/body-parsers/body-parser.json.ts (hunk 1 above)
Comment (bajtos) [inline packages/rest/src/body-parsers/body-parser.json.ts]: I think this is not the right place for rejecting array values for object-only bodies. What if the request body was sent as urlencoded formdata (or XML in the future when we add support for parsing XML request)? We also don't know what's the allowed body schema at this point.

I would expect this additional validation to happen inside [request-body.validator.ts](https://github.com/strongloop/loopback-next/blob/cf97cd9a9b8ea1ff7334de7ffb7bb870580486e2/packages/rest/src/validation/request-body.validator.ts).

Note that we already have logic to validate that non-body parameters are valid objects, see here:

https://github.com/strongloop/loopback-next/blob/cf97cd9a9b8ea1ff7334de7ffb7bb870580486e2/packages/rest/src/coercion/coerce-parameter.ts#L161-L174

Can we perhaps reuse this logic for request bodies too?

TBH, I am surprised that AJV is not handling this edge case for us and allows array values when the schema describes an object 🤔

**Comment by raymondfeng** [comment]
> I disagree. IMO, if I decorate my controller method with OpenAPI spec saying that data argument allows only objects, then LoopBack's REST layer must ensure my controller method never receives array instances.

I was confused by JSON object/array vs. JavaScript Object/Array. For JSON schema validation, `object` type should not allow JSON array.

**Comment by emonddr** [comment]
@achrinza , any update on this PR? Thank you :)

**Comment by achrinza** [comment]
@emonddr Sorry for the delay; just a bit busy lately with the current events. I'm still working on it, though I'm fine if someone else wants to take over.

**Comment by emonddr** [comment]
> @emonddr Sorry for the delay; just a bit busy lately with the current events. I'm still working on it, though I'm fine if someone else wants to take over.

Understandable, just trying to keep the discussion active :)

**Comment by dhmlau** [comment]
We just switch the contribution method from CLA to DCO, making your contribution easier in the future. Please sign the commits with DCO by amending your commit messages with `-s` flag and push the changes again. If you're using the command line, you can:

```
git commit --amend -s
git push --force-with-lease
```

Please refer to [this docs page](https://loopback.io/doc/en/contrib/code-contrib.html#developer-certificate-of-origin-dco) for details. Thanks!

**Pull request closed without merging**